### PR TITLE
Better handling of optional virtual files (e.g., shading in Figure.grdimage)

### PR DESCRIPTION
**Pull request by seisman**
### Background

Some GMT modules (e.g., `grdimage` and `grdview`) can accept a primary input file and an optional input file. Here, I'll take `grdimage` as an example. 

`grdimage` can accept an input grid file and an optional intensity grid for shading (`-I` option). Thus, in PyGMT, `Figure.grdimage`'s `shading` parameter can be given in many different ways:

1. `shading=None`: meaning no shading
2. `shading=True` or `shading=False`: shading or no shading
3. `shading=0.5`: a constant intensity to apply everywhere
4. `shading="+a30+nt0.8"`: derive the intensity from the primary input grid
5. `shading="intensity.nc"`: an intensity grid file
6. `shaidng="intensity.nc+a30+nt0.8"`: an intensity grid file plus more parameters
7. `shading=xrgrid` in which `xrgrid` is an xarray.DataArray object
8. more?

For cases 1-6, the `shading` parameter can be directly processed by the `build_arg_string` function, but for case 7, we need to create a virtual file for the xarray.DataArray grid.

### Motivations

Since the virtual file for shading depends on the value and data type of the `shading` parameter, thus we have to use [contextlib.ExitStack](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack), which is designed to make it easy to work with optional context manager. However, using `contextlib.ExitStack` still makes the codes complicated and also make it hard to reuse the codes:

https://github.com/GenericMappingTools/pygmt/blob/d2b0c39b075c41f55f67f1a9daa94365f3c6fce1/pygmt/src/grdimage.py#L182-L195

The main goal of this PR is to simplify the above codes to:
```python
    with Session() as lib:
        file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
        shade_context = lib.virtualfile_from_data(check_kind="raster", data=kwargs.get("I"))
        with file_context as fname, shade_context as shade_fname:
            kwargs["I"] = shade_fname
            lib.call_module(
                module="grdimage", args=build_arg_string(kwargs, infile=fname)
            )
```
which is more compact and more intuitive to read.

### Implementations

To achieve this, the `virtualfile_from_data` context manager must work for all cases. It's actually very simple. If `data` is not an xarray.DataArray object (cases 1-6), `virtualfile_from_data` should simply return a dummy context manager which does nothing but yield the argument passed to it (see the [`dummy_context`](https://github.com/GenericMappingTools/pygmt/blob/d2b0c39b075c41f55f67f1a9daa94365f3c6fce1/pygmt/helpers/utils.py#L96) function but also see https://github.com/GenericMappingTools/pygmt/pull/2491). This is done by changing the `data_kind` function to take arguments like None, True, False, 0.5, `"+a30+nt0.8"`, `"intensity.nc+a30+nt0.8"` as a "file" (see the changes in the `data_kind` function).

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
Currently, all the `grdimage` tests still pass, which means the new codes work, but some other tests (e.g., most `plot` tests) fail likely caused by changes in `data_kind`. So need to more careful when changing `data_kind`.

File: pygmt/helpers/utils.py
Comment (weiji14) [inline pygmt/helpers/utils.py:47]: Currently the function has `required_z` and `optional_data` parameters. Should we use `required_z` and `required_data` instead? Or `optional_z` and `optional_data`?

File: pygmt/helpers/utils.py
Comment (weiji14) [inline pygmt/helpers/utils.py:74]: Somewhat off-topic, but thinking about whether we need a doctest for the `required_z` parameter here instead of just at https://github.com/GenericMappingTools/pygmt/blob/59a203579a2b896b0d721fa88b65d7b97a5115f3/pygmt/tests/test_clib.py#L427-L453

File: pygmt/helpers/utils.py
Comment (weiji14) [inline pygmt/helpers/utils.py:79]: Kinda wishing that we can use switch statements here, but that's only Python 3.10 and above :upside_down_face: 

- https://stackoverflow.com/questions/11479816/what-is-the-python-equivalent-for-a-case-switch-statement
- https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching
- https://peps.python.org/pep-0634/

File: pygmt/helpers/utils.py
Comment (weiji14) [inline pygmt/helpers/utils.py:98]: So for `shading="+a30+nt0.8"` (a str type argument), the returned kind will be `file` instead of `null`. Technically both `file` and `null` returns a nullcontext from `virtualfile_from_data` so it shouldn't really matter, but it makes me wonder if we should combine the `file` and `null` kind, or be a bit more explicit somehow, otherwise this `data_kind` function will get confusing to new maintainers.

**Review by weiji14 — COMMENTED** [review]
A few comments for now, still need to wrap my head around the main nested if-then statement in the `data_kind` function.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py:79]: Yes, we can revisit many if-else clauses after we bump to Python>=3.10.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py:47]: Actually, I'm thinking about if we should remove `required_z` and use `required_ncols` instead. For example, in `Figure.plot()`, sometimes we need more than three columns (e.g., x, y, size, z_for_color).

**Comment by seisman** [comment]
> still need to wrap my head around the main nested if-then statement in the `data_kind` function.

I'm quite satisfied with the changes in `grdimage.py` and `grdview.py`, but I also find that codes in `data_kind` is not that easy to understand. So better ideas about `data_kind` are welcomed.

File: pygmt/helpers/utils.py
Comment (weiji14) [reply pygmt/helpers/utils.py:47]: Right, for #1076/#1117 you mean? Should probably do that in a separate PR.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py:98]: > So for `shading="+a30+nt0.8"` (a str type argument), the returned kind will be `file` instead of `null`.

Yes, maybe we can rename `file` to `file_or_arg`?

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py:47]: What are the valid values in the list? `color`, `size`, `symbol` and more? I think it will make the codes more complicated.

File: pygmt/src/grdview.py
Comment (weiji14) [inline pygmt/src/grdview.py:167]: Slightly relevant, how would we handle RGB xarray.DataArray inputs to drapegrid `-G` as mentioned at #1467? Currently [GMT 6.4.0](https://docs.generic-mapping-tools.org/6.4/grdview.html#g) requires passing it in as `-Ggrd_r -Ggrd_g -Ggrd_b`, but I'm wondering if we should only support just a single `xarray.DataArray` of shape (3, Y, X) instead.

File: pygmt/helpers/utils.py
Comment (weiji14) [reply pygmt/helpers/utils.py:98]: Naming is hard :upside_down_face: Yeah, let's go with `file_or_arg`.

File: pygmt/src/grdview.py
Comment (seisman) [reply pygmt/src/grdview.py:167]: > we should only support just a single `xarray.DataArray` of shape (3, Y, X) instead.

Sounds good to me.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py]: Done in 9a43dc8edab2b5b0f907b84e0ff740cda2c4c80e.

File: pygmt/helpers/utils.py
Comment (weiji14) [inline pygmt/helpers/utils.py:79]: > but I also find that codes in data_kind is not that easy to understand. So better ideas about data_kind are welcomed.

How about we split the function into two? A `_validate_data_input` for the first half that checks that there's no missing or invalid data, and then the second half can stay in `data_kind` and that checks the instance type of the dataset.

Comment (seisman) [reply pygmt/helpers/utils.py:79]: Sounds a good idea. Will do later.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py]: > Currently the function has `required_z` and `optional_data` parameters. Should we use `required_z` and `required_data` instead? Or `optional_z` and `optional_data`?

Renamed `optional_data` to `required_data` in 932d07d5abea788b3ca4aa0a526fe124d01f3020.

**Review by weiji14 — COMMENTED** [review]
Some documentation fixes

File: pygmt/helpers/utils.py
Comment (seisman) [inline pygmt/helpers/utils.py:66]: The logical here is still wrong, leading to ~50 failures.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py]: > So for `shading="+a30+nt0.8"` (a str type argument), the returned kind will be `file` instead of `null`.

Technically speaking, `"+a30+nt0.8"` is a valid file name, so a `file` kind still makes sense. 

The `data_kind` function is also being used in many module wrappers. For example, in `legend`,

https://github.com/GenericMappingTools/pygmt/blob/6764554c84e9b4c328b3b41d60c8337fa0f18373/pygmt/src/legend.py#L78-L83

`specfile` can be either `None` or a "file" (string type), but can't be in bool/int/float type. Thus, instead of having a single `file_or_arg` kind, we may need to have more kinds like `file` (str or PurePath), `null` (None), `bool` (True or False), `value` (int/float).

Comment (weiji14) [reply pygmt/helpers/utils.py]: Ok, maybe just keeping `"file"` kind is good rather than `"file_or_arg"`, and we can add the `"null"` kind back? Not sure if adding separate kinds for `bool` or `value` is needed, since all of these would just return a nullcontext.

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py]: Done in 73c31e214

File: pygmt/helpers/utils.py
Comment (seisman) [reply pygmt/helpers/utils.py]: Have two different kinds `file` and `arg` in commit 5faa05dc27b458e7ba675e02704faefcf05d9086

**Comment by seisman** [comment]
This PR is now completed and ready for review.

File: pygmt/clib/session.py
Comment (yvonnefroehlich) [inline pygmt/clib/session.py:1478]: Following the [Code Style](https://www.pygmt.org/dev/contributing.html#code-style), no underscore should be used for words bridged only by consonants. However, already `required_z` and `check_kind` contain underscores...
```suggestion
        required_z=False,
        requireddata=True,
```

Comment (seisman) [reply pygmt/clib/session.py:1478]: These are internal functions for developers/maintainers. So maybe we don't have to strictly follow the code style?

Comment (weiji14) [reply pygmt/clib/session.py:1478]: That code style is meant for GMT aliases, so we can use `required_data` here.

File: pygmt/clib/session.py
Comment (weiji14) [inline pygmt/clib/session.py:1547]: Debating on whether to remove this else argument. Technically this `lib.virtualfile_from_data` function is for internal use only, but the function name is not marked as private with a leading underscore (`_virtualfile_from_data`). Should we assume that others are using this and might get an error with the default `check_kind=None`? Or add a test for `check_kind=None` so that there is code coverage?

File: pygmt/helpers/utils.py
Comment (weiji14) [inline pygmt/helpers/utils.py:139]: Similar suggestion to above, feel free to reword.
```suggestion
        Set to True when 'data' is required, or False when dealing with
        optional virtual files. [Default is True].
```

File: pygmt/clib/session.py
Comment (weiji14) [inline pygmt/clib/session.py:1543]: Suggested refactoring. Makes it a little clearer that raster includes "grid", and vector includes "matrix", "vectors", "geojson".

```suggestion
        valid_kinds = ("file", "arg") if required_data is False else ("file",)
        if check_kind == "raster":
            valid_kinds += ("grid",)
        elif check_kind == "vector":
            valid_kinds += ("matrix", "vectors", "geojson")
```

File: pygmt/clib/session.py
Comment (weiji14) [inline pygmt/clib/session.py:1549]:
```suggestion

```

**Review by weiji14 — COMMENTED** [review]
Nice work getting the if-then logic correct! Still a bit hard to follow in some parts, but I think this is the best that we can do. Just some more minor comments/suggestions and should be close to merge.

File: pygmt/clib/session.py
Comment (seisman) [reply pygmt/clib/session.py]: The docstring says:
```
Used to validate the type of data that can be passed in. Choose     
from 'raster', 'vector', or None. Default is None (no validation).
```
so we should not check the data type if check_kind is None. See 9b31e230fa0f4640e3a0a5bd12f5aad1c58936c0 for the changes.

Comment (weiji14) [reply pygmt/clib/session.py]: Sounds good. Could you also change the docstring for the expected type at L1489 to `check_kind : str or None` please.

Comment (seisman) [reply pygmt/clib/session.py]: Done in 0ed6248bb76bda675c5d87c51b123b48860b7213

**Review by weiji14 — APPROVED** [review]
Excellent, thanks @seisman! Will be easier to work on #2590 after this PR is merged :smile:

File: pygmt/clib/session.py
Comment (yvonnefroehlich) [reply pygmt/clib/session.py:1478]: OK.

**Pull request merged**
